cpp/src/preprocessing/spectral/spectral_embedding.cu

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cuvs/neighbors/brute_force.hpp>
+#include <cuvs/preprocessing/spectral_embedding.hpp>
+
+#include <raft/core/device_coo_matrix.hpp>
+#include <raft/core/device_mdspan.hpp>
+#include <raft/core/handle.hpp>
+#include <raft/core/resources.hpp>
+#include <raft/linalg/matrix_vector_op.cuh>
+#include <raft/matrix/gather.cuh>
+#include <raft/matrix/init.cuh>
+#include <raft/sparse/coo.hpp>
+#include <raft/sparse/linalg/laplacian.cuh>
+#include <raft/sparse/linalg/symmetrize.cuh>
+#include <raft/sparse/op/filter.cuh>
+#include <raft/sparse/solver/lanczos.cuh>
+#include <raft/sparse/solver/lanczos_types.hpp>
+#include <raft/util/cudart_utils.hpp>
+#include <raft/util/integer_utils.hpp>
+
+#include <thrust/sequence.h>
+#include <thrust/tabulate.h>
+
+namespace cuvs::preprocessing::spectral_embedding {
+
+void transform(raft::resources const& handle,
+               params spectral_embedding_config,
+               raft::device_matrix_view<float, int, raft::row_major> dataset,
+               raft::device_matrix_view<float, int, raft::col_major> embedding)
+{
+  const int n_samples   = dataset.extent(0);
+  const int n_features  = dataset.extent(1);
+  const int k_search    = spectral_embedding_config.n_neighbors;
+  const bool drop_first = spectral_embedding_config.drop_first;
+  const size_t nnz      = n_samples * k_search;
+
+  auto stream = raft::resource::get_cuda_stream(handle);
+
+  cuvs::neighbors::brute_force::search_params search_params;
+  cuvs::neighbors::brute_force::index_params index_params;
+  index_params.metric = cuvs::distance::DistanceType::L2SqrtExpanded;
+
+  auto d_indices   = raft::make_device_matrix<int64_t>(handle, n_samples, k_search);
+  auto d_distances = raft::make_device_matrix<float>(handle, n_samples, k_search);
+
+  auto index =
+    cuvs::neighbors::brute_force::build(handle, index_params, raft::make_const_mdspan(dataset));
```

Comment:
Please pull this out into a "compute_nearest_neighbors" function. This will allow us to eventually swap the brute-force version out for an ANN with minimal changes.